### PR TITLE
tests for RampageReqRampageRequestHandleruestHandler

### PR DESCRIPTION
**Pull request by bennija**
this does not work, so how do I use those factories?

### CODE REVIEW

File: test/RampageRequestHandlerTest.php [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+        $this->assertSame("???","???");
+    }
+
+    public function testHandle()
+    {
+
+        $responseFactory = new ResponseFactory;
+        $streamFactory = new StreamFactory;
+        $handler = new RampageRequestHandler($responseFactory, $streamFactory);
+        $serverRequestMock = $this->createMock(ServerRequestInterface::class);
```
Comment (midahp) [inline test/RampageRequestHandlerTest.php:58]: I would not use a mock here. Use the request factory.

File: test/RampageRequestHandlerTest.php [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+    }
+
+    public function testSetPayloadHandler()
+    {
+
+        $responseFactory = new ResponseFactory;
+        $streamFactory = new StreamFactory;
+        $handler = new RampageRequestHandler($responseFactory, $streamFactory);
+        $requestHandlerMock = $this->createMock(RequestHandlerInterface::class);
+        $handler->setPayloadHandler($requestHandlerMock);
```
Comment (midahp) [inline test/RampageRequestHandlerTest.php:47]: The `handle` method of payloadHandler will be called, so try to mock that method.

File: test/RampageRequestHandlerTest.php [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+        $streamFactory = new StreamFactory;
+        $handler = new RampageRequestHandler($responseFactory, $streamFactory);
+        $requestFactory = new RequestFactory();
+        //! $request = $requestFactory->createServerRequest('GET', 'https://www.horde.org');
+            //!TypeError: Argument 1 passed to Horde\Http\Server\RampageRequestHandler::addMiddleware()
+                //! must be an instance of Psr\Http\Server\MiddlewareInterface, instance of Horde\Http\ServerRequest given.
+        //Todo instead with mock, but how to use responseFactory in that context
+        //$request = new Mock(???);
+        $middlewareMock = $this->createMock(MiddlewareInterface::class);
+        $handler->addMiddleware($middlewareMock);
```
Comment (midahp) [inline test/RampageRequestHandlerTest.php]: This is fine. But you should add like three middlewares and then check if they are returned by `nextMiddleware` FIFO order.

**Review by midahp — CHANGES_REQUESTED** [review]
In addition to the mentioned issues add tests for `nextMiddleware` and make sure you add test for all code paths in `handle`.

File: test/RampageRequestHandlerTest.php [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+        $middlewareMock1 = $this->createMock(MiddlewareInterface::class);
+        $middlewareMock1->n = 1;
+        $middlewareMock2 = $this->createMock(MiddlewareInterface::class);
+        $middlewareMock2->n = 2;
+        $middlewareMock3 = $this->createMock(MiddlewareInterface::class);
+        $middlewareMock3->n = 3;
+        $handler->addMiddleware($middlewareMock1);
+        $handler->addMiddleware($middlewareMock2);
+        $handler->addMiddleware($middlewareMock3);
+        $firstInNumber = $handler->nextMiddleware()->n;
```
Comment (midahp) [inline test/RampageRequestHandlerTest.php:30]: Check the other middlewares as well. Also test if it returns `null` after the middleware stack has been exhausted.

File: test/RampageRequestHandlerTest.php [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+        $handler = new RampageRequestHandler($responseFactory, $streamFactory);
+        $serverRequestMock = $this->createMock(ServerRequestInterface::class);
+        $requestHandlerMock = $this->createMock(RequestHandlerInterface::class);
+        $handler->setPayloadHandler($requestHandlerMock);
+        $requestHandlerMock->method('handle')->willReturn($responseFactory->createResponse());
+        $handled = $handler->handle($serverRequestMock);
+        $this->assertSame(200,$handled->getStatusCode());
+    }
+
+    public function testHandle()
```
Comment (midahp) [inline test/RampageRequestHandlerTest.php]: You still need to test the handle method when middlewares are set.

File: test/RampageRequestHandlerTest.php [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+    {
+
+        $responseFactory = new ResponseFactory;
+        $streamFactory = new StreamFactory;
+        $requestFactory = new RequestFactory;
+        $handler = new RampageRequestHandler($responseFactory, $streamFactory);
+        $handled = $handler->handle($requestFactory->createServerRequest('GET','www.test.de'));
+        $this->assertSame(404,$handled->getStatusCode());
```
Comment (midahp) [inline test/RampageRequestHandlerTest.php]: Its probably better to check for the reason phrase as well here.

File: test/RampageRequestHandlerTest.php (hunk 5 above)
Comment (bennija) [reply test/RampageRequestHandlerTest.php]: Im not sure if the new test does its job.

File: test/RampageRequestHandlerTest.php [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+        $handler = new RampageRequestHandler($responseFactory, $streamFactory);
+        $serverRequestMock = $this->createMock(ServerRequestInterface::class);
+        $requestHandlerMock = $this->createMock(RequestHandlerInterface::class);
+        $handler->setPayloadHandler($requestHandlerMock);
+        $requestHandlerMock->method('handle')->willReturn($responseFactory->createResponse());
+        $handled = $handler->handle($serverRequestMock);
+        $this->assertSame(200,$handled->getStatusCode());
+    }
+
+    public function testHandleWithSetMiddelware()
```
Comment (midahp) [inline test/RampageRequestHandlerTest.php:53]: The payload handler is not really needed here.
You want to make sure the middleware is used by the handle method.
I think the best way to do this would be to make the middleware return a specific response and then check for that response at the end.
